File: python/packages/botas/src/botas/state/file_storage.py
"""File-based storage implementation using JSON files on disk."""

from __future__ import annotations

import asyncio
import json
from pathlib import Path
from typing import Union
from urllib.parse import quote


class FileStorage:
    """JSON file-based storage for bot state.

    Stores each state key as a separate JSON file in a configurable directory.
    Keys are sanitized for filesystem safety. Suitable for single-instance
    deployments where simple persistence is needed.

    **Not suitable for multi-instance deployments** — no locking or concurrency control.

    Example::

        from botas.state import FileStorage

        storage = FileStorage("./bot-state")
        await storage.write({"conversation/123": {"count": 5}})
        data = await storage.read(["conversation/123"])
        # data = {"conversation/123": {"count": 5}}

    Args:
        root_path: Root directory for state files. Defaults to ``"./bot-state"``.
    """

    def __init__(self, root_path: Union[str, Path] = "./bot-state") -> None:
        """Initialize file storage with a root directory.

        Args:
            root_path: Root directory for state files. Defaults to ``"./bot-state"``.
        """
        self._root = Path(root_path)

    def _sanitize_key(self, key: str) -> str:
        """Sanitize a storage key for filesystem safety.

        Uses URL percent-encoding with no safe characters, ensuring
        cross-platform filesystem compatibility.

        Args:
            key: Raw storage key (e.g., "msteams/bot-id/conversations/conv-id").

        Returns:
            Filesystem-safe encoded key.
        """
        return quote(key, safe="")

    def _key_to_path(self, key: str) -> Path:
        """Convert a storage key to a file path.

        Args:
            key: Storage key.

        Returns:
            Absolute path to the JSON file for this key.
        """
        sanitized = self._sanitize_key(key)
        return self._root / f"{sanitized}.json"

    async def read(self, keys: list[str]) -> dict[str, object]:
        """Read items from storage.

        Args:
            keys: Keys to read.

        Returns:
            Dictionary of key-value pairs that exist in storage.
            Missing keys are omitted from the result.
        """
        result: dict[str, object] = {}
        for key in keys:
            path = self._key_to_path(key)
            try:
                # Use asyncio.to_thread to avoid blocking the event loop
                content = await asyncio.to_thread(path.read_text, encoding="utf-8")
                result[key] = json.loads(content)
            except FileNotFoundError:
                # Missing file is not an error — just omit from result
                pass
        return result

    async def write(self, changes: dict[str, object]) -> None:
        """Write items to storage.

        Creates parent directories if they don't exist.

        Args:
            changes: Dictionary of key-value pairs to write.
        """
        # Ensure root directory exists
        await asyncio.to_thread(self._root.mkdir, parents=True, exist_ok=True)

        for key, value in changes.items():
            path = self._key_to_path(key)
            content = json.dumps(value, ensure_ascii=False, indent=2)
            await asyncio.to_thread(path.write_text, content, encoding="utf-8")

    async def delete(self, keys: list[str]) -> None:
        """Delete items from storage.

        Args:
            keys: Keys to delete. Idempotent — no error if key doesn't exist.
        """
        for key in keys:
            path = self._key_to_path(key)
            try:
                await asyncio.to_thread(path.unlink)
            except FileNotFoundError:
                # Idempotent — no error if file doesn't exist
                pass

File: node/packages/botas-core/src/state/file-storage.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { mkdir, readFile, writeFile, unlink } from 'node:fs/promises'
import { join, dirname } from 'node:path'
import type { Storage } from './storage.js'

/**
 * File-based storage implementation for simple persistence in single-instance deployments.
 *
 * Stores each key as a separate JSON file in the configured directory.
 * **Not suitable for multi-instance deployments** — no locking or concurrency protection.
 *
 * @example
 * ```ts
 * // Default directory: './bot-state'
 * const storage = new FileStorage()
 *
 * // Custom directory
 * const storage = new FileStorage('./data/state')
 * bot.useState(storage)
 * ```
 */
export class FileStorage implements Storage {
  private readonly rootDir: string

  /**
   * Create a new FileStorage instance.
   * @param rootDir - Root directory for state files. Defaults to './bot-state'.
   */
  constructor(rootDir: string = './bot-state') {
    this.rootDir = rootDir
  }

  /**
   * Read items from file storage.
   * @param keys - Keys to read.
   * @returns Dictionary of key-value pairs that exist in storage.
   */
  async read(keys: string[]): Promise<Record<string, unknown>> {
    const result: Record<string, unknown> = {}
    
    await Promise.all(
      keys.map(async (key) => {
        const filePath = this.getFilePath(key)
        try {
          const content = await readFile(filePath, 'utf-8')
          result[key] = JSON.parse(content)
        } catch (err) {
          // File doesn't exist or invalid JSON - return nothing for this key
          // This matches the spec: "read() returns null/undefined when file missing"
        }
      })
    )

    return result
  }

  /**
   * Write items to file storage.
   * @param changes - Dictionary of key-value pairs to write.
   */
  async write(changes: Record<string, unknown>): Promise<void> {
    await Promise.all(
      Object.entries(changes).map(async ([key, value]) => {
        const filePath = this.getFilePath(key)
        const dir = dirname(filePath)
        
        // Create parent directories if needed
        await mkdir(dir, { recursive: true })
        
        // Write JSON to file
        const content = JSON.stringify(value, null, 2)
        await writeFile(filePath, content, 'utf-8')
      })
    )
  }

  /**
   * Delete items from file storage.
   * @param keys - Keys to delete.
   */
  async delete(keys: string[]): Promise<void> {
    await Promise.all(
      keys.map(async (key) => {
        const filePath = this.getFilePath(key)
        try {
          await unlink(filePath)
        } catch (err) {
          // Idempotent - ignore if file doesn't exist (ENOENT)
        }
      })
    )
  }

  /**
   * Convert a storage key to a safe file path.
   * Uses encodeURIComponent to sanitize keys with separators or invalid chars.
   */
  private getFilePath(key: string): string {
    // Encode key to make it path-safe (matches .NET encoding for cross-language parity)
    const sanitized = encodeURIComponent(key)
    return join(this.rootDir, `${sanitized}.json`)
  }
}
